{% extends "layout.html" %}

{% block title %}Sample: {{ sample.sample_id }}{% endblock %}

{% block head %}
    {{ super() }}
    <script type="text/javascript" src="{{ url_for('static', filename='js/mm10-data.js') }}"></script>
    <!-- script to save a D3 plot (svg) as a png. MIT license - author Eric Shull (github.com/exupero/saveSvgAsPng) -->
    <script type="text/javascript" src="{{ url_for('static', filename='js/saveSvgAsPng.js') }}"></script>
    <script>

        function arraysEq(arr1, arr2) {
            if(arr1.length !== arr2.length) {
                return false;
            } else {
                return arr1.every(function(elem1, i) {
                    return elem1 === arr2[i];
                });
            }
        }

        var strainMap = {{ strain_map|tojson }};
        var strainNames = $.map(strainMap, function(_, strainName) {return strainName});
        strainNames.sort();
        var strainIndexMap = {};
        strainNames.forEach(function(strainName, i) {
            strainIndexMap[strainName] = i;
        });

        var sampleID = {{ sample.sample_id|tojson }};
        {% if sample.strain_id %}
        var strainID = {{ sample.strain_id|tojson }}
        {% else %}
        var strainID = ""
        {% endif %}
        var sex = {{ sample.sex|tojson }};
        if(sex === null) {
            sex = 'unknown';
        }

        var allSds = {{ all_sds|tojson }};
        var allTags = {{ all_tags|tojson }};
        var allOwners = {{ all_owners|tojson }};
        var sample = {{ sample|tojson }};
        var allEngTgts = {{ all_eng_tgts|tojson }};
        var haplotypePollingIntervalMillisecs = 2500;
        var comparisonStatus = true;

        $(function() {

            var editOwnerInput = $('#edit-owner-input');

            $('#hap-candidate-checkbox').prop('checked', sample.haplotype_candidate);

            editOwnerInput.tokenfield({
                autocomplete: {
                    source: allOwners.map(function(owner) {
                        return {
                            label: owner,
                            value: owner
                        }
                    }),
                    delay: 100,
                    autoFocus: true,
                    //minWidth isn't working so setting inline style
                    minWidth: 300,
                    minLength: 1
                },
                showAutocompleteOnFocus: true,
                createTokensOnBlur: true
            });

            //export karyotype plot to png
            $('#export-png').click(function () {
                var svgElement = document.getElementById('hapkaryoplot-svg');
                saveSvgAsPng(svgElement, "{{ sample.sample_id | urlencode }}.png");
            });

            /**
             * loads the comparison select with the other samples from the same
             * standard_designation
             */
            function getComparisonSamples () {
                var samplesURL = "/standard-designation.json/" + encURIComp(sample.standard_designation);
                var comparisonSelect = $("#comparison-sample");
                $.getJSON(samplesURL, function(data) {
                    // if there is data and there's more than 1 sample in the standard designation (SD)
                    if (typeof data.samples !== "undefined" && data.samples.length > 1) {
                        var comp_samples = [];
                        data.samples.forEach(function(sample) {
                            // only load samples that aren't the current
                            if (sample.sample_id !== sampleID) {
                                var option = new Option(sample.sample_id, sample._id);
                                comparisonSelect.append(option);
                                comp_samples.push(sample);
                            }
                        });

                        sample.comp_samples = comp_samples;
                    }
                    else {
                        // if there's no comparison, hide all of the comparison sample-related elements
                        comparisonStatus = false;
                        d3.selectAll(".comparison-only").remove();
                    }

                }).success(function() {
                    // only if there are other samples in the SD
                    if(comparisonStatus) {
                        var sampleSelected = comparisonSelect.children("option").filter(":selected");
                        sample.current_comp = sampleSelected.val();
                        updateComparisonSVG(sample.current_comp, 2);

                        // if an interval has been chosen, use it, else default to chr 1
                        if (interval) {
                            getSNPData(sample.current_comp, interval.chr, hapIntervalCompPlot);
                        }
                        else {
                            getSNPData(sample.current_comp, 1, hapIntervalCompPlot);
                        }
                        var labelText = sampleSelected.text();
                        $("#comp-label").text("Sample: " + labelText);

                        // if the previous chosen sample resulted in no data, remove the overlay
                        d3.select("[id=no-data-overlay]").remove();
                    }
                });
            }

            getComparisonSamples();

            // load data when a comparison sample is chosen
            $("#comparison-sample").on("change", function() {
                sample.current_comp = $(this).val();

                var labelText = $(this).children("option").filter(":selected").text();
                $("#comp-label").text("Sample: " + labelText);

                // if the previous chosen sample resulted in no data, remove the overlay
                d3.select("[class=no-data-overlay]").remove();

                // update the comparison svg and only attempt data retrieval twice
                updateComparisonSVG(sample.current_comp, 2);

                // if an interval has been chosen, use it, else default to chr 1
                if(interval) {
                    getSNPData(sample.current_comp, interval.chr, hapIntervalCompPlot);
                }
                else {
                    getSNPData(sample.current_comp, 1, hapIntervalCompPlot);
                }
            });

            // when carousel previous button is pressed, get the previous sample
            // carousel loops so if current sample is first, previous gets last sample
            $("#comp-prev").on("click", function() {
                var index;

                for (var i = 0; i < sample.comp_samples.length; i++) {
                    if (sample.current_comp === sample.comp_samples[i]._id) {
                        index = i;
                    }
                }

                if(index === 0) {
                    var lastIndex = sample.comp_samples.length - 1;
                    $("#comparison-sample").val(sample.comp_samples[lastIndex]._id).trigger("change");
                }
                else {
                    $("#comparison-sample").val(sample.comp_samples[index-1]._id).trigger("change");
                }
            });

            // when carousel next button is pressed, get the next sample
            // carousel loops so if current sample is last, previous gets first sample
            $("#comp-next").on("click", function() {
                var index;

                for (var i = 0; i < sample.comp_samples.length; i++) {
                    if (sample.current_comp === sample.comp_samples[i]._id) {
                        index = i;
                    }
                }

                var lastIndex = sample.comp_samples.length - 1;
                if(index === lastIndex) {
                    $("#comparison-sample").val(sample.comp_samples[0]._id).trigger("change");
                }
                else {
                    $("#comparison-sample").val(sample.comp_samples[index+1]._id).trigger("change");
                }
            });

            var candidateHaplotypeTable = $('#candidate-haplotype-table');
            var candidateHaplotypeTableOverlay = new ElemOverlay(
                    candidateHaplotypeTable,
                    $('<span><span class="fa fa-refresh fa-spin" aria-hidden="true"></span> Recalculating Candidate Haplotypes</span>'));
            var candidateHaplotypeTableBody = $('#candidate-haplotype-table-body');
            var findCandidateHapsReq = null;
            var findCandidateHapsWaiting = false;

            function findCandidateHaps() {
                if(interval === null) {
                    if(findCandidateHapsReq !== null) {
                        findCandidateHapsReq.abort();
                        findCandidateHapsReq = null;
                        findCandidateHapsWaiting = false;
                    }
                } else {
                    if(findCandidateHapsReq !== null) {
                        // we have to wait until this request finishes before we
                        // launch another to avoid overloading the server
                        findCandidateHapsWaiting = true;
                    } else {
                        candidateHaplotypeTableOverlay.overlayActive(true);
                        var startPos = interval.startPos;
                        var stopPos = interval.startPos + interval.size - 1;
                        if(startPos < 0) {
                            startPos = 0;
                        }
                        if(stopPos < startPos) {
                            stopPos = startPos;
                        }
                        var candHapURL =
                                '../best-haplotype-candidates/{{ sample._id }}/chr' +
                                interval.chr + '-' + startPos + '-' + stopPos + '.json';
                        $.getJSON(candHapURL, function(data) {
                            candidateHaplotypeTableBody.empty();
                            data.best_candidates.forEach(function(currCandidate) {
                                function makeHapLink(strainName) {
                                    var strainIdx = strainNames.indexOf(strainName);

                                    return ($('<a href="/standard-designation/' + encURIComp(strainName)
                                        + '.html" class="hap hap' + strainIdx + '">'
                                        + '<span id="sample-color-icon" class="fa fa-square" style="color: '
                                        + strainMap[strainName].color + ';"></span>'
                                        + escapeHtml(strainName) + '</a>'))
                                }

                                var tr = $(document.createElement('tr'));

                                var labelsTD = $(document.createElement('td'));
                                labelsTD.append(makeHapLink(currCandidate.haplotype_1));
                                if(currCandidate.haplotype_1 !== currCandidate.haplotype_2) {
                                    labelsTD.append($(document.createTextNode(' | ')));
                                    labelsTD.append(makeHapLink(currCandidate.haplotype_2));
                                }
                                tr.append(labelsTD);

                                var likelihoodTD = $(document.createElement('td'));
                                likelihoodTD.text(currCandidate.neg_log_likelihood.toFixed(2));
                                tr.append(likelihoodTD);

                                candidateHaplotypeTableBody.append(tr);
                            });

                            updateHaplotypeLinkHoverListeners();

                            // see if we have a pending find
                            findCandidateHapsReq = null;
                            if(findCandidateHapsWaiting) {
                                findCandidateHapsWaiting = false;
                                findCandidateHaps();
                            }
                        }).fail(function() {
                            showErrorMessage('An error occurred while calculating candidate haplotypes.');
                        }).always(function() {
                            candidateHaplotypeTableOverlay.overlayActive(false);
                        });
                    }
                }
            }

            var gemmTable = $('#gemm-table');
            var gemmTableOverlay = new ElemOverlay(
                    gemmTable,
                    $('<span><span class="fa fa-refresh fa-spin" aria-hidden="true"></span> Calculating GEMM Probabilities</span>'));
            var gemmTableBody = $('#gemm-table-body');
            //TODO: this doesn't appear to be necessary. leaving in for now in case its needed later
            //gemmTableOverlay.overlayActive(true);

            $.getJSON("{{ url_for('gemm_probs_json', mongo_id=sample._id) }}", function (data) {
                    if (data.gemm_probs.length) {
                        gemmTableBody.empty();
                        data.gemm_probs.forEach(function (gemProbPair) {
                            var gemmName = gemProbPair[0];
                            var gemmProb = gemProbPair[1];
                            var tr = $(document.createElement('tr'));

                            var labelsTD = $(document.createElement('td'));
                            labelsTD.text(gemmName);
                            tr.append(labelsTD);

                            var likelihoodTD = $(document.createElement('td'));
                            if(gemmProb === null) {
                                likelihoodTD.text('N/A');
                            } else {
                                likelihoodTD.text(gemmProb.toFixed(4));
                            }
                            tr.append(likelihoodTD);

                            gemmTableBody.append(tr);
                        });
                    }
                }).fail(function () {
                    //TODO: need to know what's going on when this fails for a non-logged in user
                    //letting this fail for now as the plot still gets rendered
                    //showErrorMessage('An error occurred while calculating GEMM probabilities.');
                    console.log('An error occurred while calculating GEMM probabilities.');
                }).always(function () {
                    gemmTableOverlay.overlayActive(false);
                });

            var interval = null;
            var lastChr = null;
            var updateIntervalButton = $('#update-interval-button');
            var clearIntervalButton = $('#clear-interval-button');
            var intervalText = $('#interval-edit');
            var gotoEnsemblLink = $('#goto-ensembl-link');
            var gotoUCSCLink = $('#goto-ucsc-link');
            gotoEnsemblLink.addClass('disabled');
            gotoUCSCLink.addClass('disabled');

            var updateIntervalDP = new DelayedPreemptable();

            function updateInterval(newInterval) {
                if(interval === null || lastChr !== newInterval.chr) {
                    // get the snp data for the plots
                    getSNPData(sample._id, newInterval.chr, hapIntervalPlot);
                    getSNPData(sample.current_comp, newInterval.chr, hapIntervalCompPlot);
                    lastChr = newInterval.chr;
                }

                interval = newInterval;

                if(newInterval === null) {
                    intervalText.val('');
                    gotoEnsemblLink.addClass('disabled');
                    gotoUCSCLink.addClass('disabled');
                } else {
                    newInterval.endPos = newInterval.startPos + newInterval.size - 1;
                    if(Math.round(newInterval.startPos / 1000000) == newInterval.startPos / 1000000 &&
                       Math.round(newInterval.endPos / 1000000) == newInterval.endPos / 1000000) {
                        intervalText.val(
                                'Chr' + newInterval.chr + ':' +
                                (newInterval.startPos / 1000000) + 'Mb-' +
                                (newInterval.endPos / 1000000) + 'Mb');
                    } else {
                        intervalText.val(
                                'Chr' + newInterval.chr + ':' +
                                Math.round(newInterval.startPos) + '-' +
                                Math.round(newInterval.endPos));
                    }

                    // update the "go to" genome browser links for the new interval
                    var startPos = interval.startPos;
                    var stopPos = interval.startPos + interval.size - 1;
                    if(startPos < 0) {
                        startPos = 0;
                    }
                    if(stopPos > startPos) {
                        var rangeStr = encodeURIComponent(interval.chr + ':' + startPos + '-' + stopPos);
                        gotoEnsemblLink.attr(
                                'href',
                                'http://ensembl.org/Mus_musculus/Location/View?r=' + rangeStr);
                        gotoUCSCLink.attr(
                                'href',
                                'https://genome.ucsc.edu/cgi-bin/hgTracks?db=mm10&position=chr' + rangeStr);
                    }
                    gotoEnsemblLink.removeClass('disabled');
                    gotoUCSCLink.removeClass('disabled');
                }

                hapKaryoPlot.zoomInterval(newInterval);
                hapIntervalPlot.zoomInterval(newInterval);
                hapIntervalCompPlot.zoomInterval(newInterval);

                // don't update haplotypes unless we have at least 1 sec of quiet time.
                updateIntervalDP.delay(findCandidateHaps, 1000);
            }

            clearIntervalButton.click(function() {
                updateInterval(null);
            });

            updateIntervalButton.click(function() {
                updateInterval(parseInterval(intervalText.val()));
            });

            var editTagsInput = $('#edit-tags-input');
            var tagLinksPanel = $('#tag-links');

            var editPosEngTgtsInput = $('#edit-pos-eng-tgt-input');
            var posEngTgtLinks = $('#pos-eng-tgt-links');
            var editNegEngTgtsInput = $('#edit-neg-eng-tgt-input');
            var negEngTgtLinks = $('#neg-eng-tgt-links');

            var sampleUpdateSpinner = $('#sample-update-spinner');
            var editSampleButton = $('#edit-sample-button');
            var saveSampleButton = $('#save-sample-button');
            var cancelSampleButton = $('#cancel-sample-button');
            var runHaplotypingButton = $('#run-haplotyping');

            var readonlyHaplotypesPanel = $('.readonly-haplotypes-panel');
            var readonlyCompHaplotypesPanel = $('.readonly-comp-haplotypes-panel');

            var editSampleIDInput = $('#edit-sampleid-input');

            var editStandardDesignationInput = $('#edit-standarddesignation-input');
            var readonlyStandardDesignationPanel = $('#readonly-standarddesignation');

            var editNotesInput = $('#edit-notes-input');
            var readonlyNotesPanel = $('#readonly-notes');

            var sampleIsPublicCheckbox = $('#sample-is-public-checkbox');
            var readonlySampleVisibility = $('#readonly-sample-visibility');

            var editStrainIdInput = $('#edit-strain-id-input');
            var readonlyStrainId = $('#readonly-strain-id');

            var editSexSelect = $('#edit-sex-select');
            var readonlySex = $('#readonly-sex');

            function updateSex(newSex) {
                sex = newSex;

                var sexLower = newSex.toLowerCase();
                if(['female', 'male', 'unknown'].indexOf(sexLower) === -1) {
                    editSexSelect.val('unknown');
                } else {
                    editSexSelect.val(sexLower);
                }
                readonlySex.text(newSex);
            }

            function updateSampleStrainId(newSampleStrainId) {
                readonlyStrainId.html(newSampleStrainId);
            }

            function updateSampleIsPublic(newSampleIsPublic) {
                sample.is_public = newSampleIsPublic;

                sampleIsPublicCheckbox.prop('checked', newSampleIsPublic);
                readonlySampleVisibility.text(newSampleIsPublic ? 'Publicly Visible' : 'Requires Login');
            }

            editStandardDesignationInput.tokenfield({
                autocomplete: {
                    source: allSds.map(function(stdes) {
                        return {
                            label: stdes,
                            value: stdes
                        }
                    }),
                    delay: 100,
                    autoFocus: true,
                    minLength: 1
                },
                showAutocompleteOnFocus: true,
                createTokensOnBlur: true
            });

            editTagsInput.tokenfield({
                autocomplete: {
                    source: allTags.map(function(tag) {
                        return {
                            label: tag,
                            value: tag
                        }
                    }),
                    delay: 100,
                    autoFocus: true,
                    minLength: 1
                },
                showAutocompleteOnFocus: true,
                createTokensOnBlur: true
            });

            function updateSampleTags(newSampleTags) {
                newSampleTags.sort(function(x, y) {
                    return x.localeCompare(y);
                });
                sample.tags = newSampleTags;
                tagLinksPanel.empty();
                // TODO replace the haplotype-strain check
                var haplotypeStrainTagPresent = false;
                newSampleTags.forEach(function(tag, i) {
                    if(tag === 'haplotype-strain') {
                        haplotypeStrainTagPresent = true;
                    }

                    if(i >= 1) {
                        tagLinksPanel.append(document.createTextNode(', '));
                    }
                    tagLinksPanel.append('<a href="../tag/' + encURIComp(tag) + '.html">' + tag + '</a>');
                });

                editTagsInput.tokenfield('setTokens', newSampleTags.map(function(tag) {
                    return {
                        label: tag,
                        value: tag
                    }
                }));
            }

            var engTgtsTokenfieldParams = {
                autocomplete: {
                    source: allEngTgts.map(function(engTgt) {
                        return {
                            label: engTgt,
                            value: engTgt
                        }
                    }),
                    delay: 100,
                    autoFocus: true,
                    minLength: 1
                },
                showAutocompleteOnFocus: true,
                createTokensOnBlur: true
            };

            function createEngTgtToken(e) {
                // before a target is generated make sure it is valid (i.e. it should be
                // in the allEngTgts list
                var value = e.attrs.value;
                return allEngTgts.indexOf(value) >= 0;
            }

            editPosEngTgtsInput.tokenfield(engTgtsTokenfieldParams);
            editPosEngTgtsInput.on('tokenfield:createtoken', createEngTgtToken);
            editNegEngTgtsInput.tokenfield(engTgtsTokenfieldParams);
            editNegEngTgtsInput.on('tokenfield:createtoken', createEngTgtToken);

            function updateCtrlEngTgts(newEngTgts, isPosCtrl) {
                newEngTgts.sort(function(x, y) {
                    return x.localeCompare(y);
                });

                var samplePropStr = null;
                var tgtInput = null;
                var linksPanel = null;
                if(isPosCtrl) {
                    samplePropStr = 'pos_ctrl_eng_tgts';
                    tgtInput = editPosEngTgtsInput;
                    linksPanel = posEngTgtLinks;
                } else {
                    samplePropStr = 'neg_ctrl_eng_tgts';
                    tgtInput = editNegEngTgtsInput;
                    linksPanel = negEngTgtLinks;
                }
                sample[samplePropStr] = newEngTgts;

                linksPanel.empty();
                if(newEngTgts.length) {
                    newEngTgts.forEach(function (tgt, i) {
                        if (i >= 1) {
                            // TODO add links here rather than text nodes once we have
                            // pages for engineered targets
                            linksPanel.append(document.createTextNode(', '));
                        }
                        linksPanel.append(document.createTextNode(tgt));
                    });
                } else {
                    linksPanel.text('None');
                }
                tgtInput.tokenfield('setTokens', newEngTgts.map(function(tgt) {
                    return {
                        label: tgt,
                        value: tgt
                    };
                }));
            }

            function updateStandardDesignation(newStandardDesignation) {
                sample.standard_designation = newStandardDesignation;

                readonlyStandardDesignationPanel.empty();
                if(newStandardDesignation) {
                    readonlyStandardDesignationPanel.append(
                            '<a href="../standard-designation/' + encURIComp(newStandardDesignation) + '.html">' +
                            newStandardDesignation + '</a>');
                }

                editStandardDesignationInput.val(newStandardDesignation);
            }

            updateStandardDesignation(sample.standard_designation);

            function updateNotes(newNotes) {
                if(!newNotes) {
                    newNotes = '';
                }
                readonlyNotesPanel.empty();
                sample.notes = newNotes;
                newNotes.split(/\n/).forEach(function(line) {
                    var p = $(document.createElement('p'));
                    p.text(line);
                    readonlyNotesPanel.append(p);
                });

                editNotesInput.val(newNotes);
            }

            updateNotes(sample.notes);

            var contribStrainTokensInput = $('#contrib-strains-tokens');

            contribStrainTokensInput.tokenfield({
                autocomplete: {
                    source: strainNames,
                    delay: 100,
                    autoFocus: true
                },
                showAutocompleteOnFocus: true,
                createTokensOnBlur: true,
                limit: {{ maximum_contributing_strain_count() }}
            });

            contribStrainTokensInput.on('tokenfield:createtoken', function(e) {
                return strainNames.indexOf(e.attrs.value) >= 0;
            });

            var strainTS = {};
            var changes = [];

            // gets all last_update timestamps for contributing strains
            function getLastContributingCandidateUpdates() {
                sample.contributing_strains.forEach(function (strain) {
                    var url = '/get-hap-candidate-ts/' + strain.replace("/", "@");
                    $.getJSON(url, function (data) {
                        if(data.last_update) {
                            strainTS[strain] = data.last_update;
                        }
                        else {
                            strainTS[strain] = null
                        }
                    });
                });
            }

            // makes bootstrap alert with links to all of the candidate strains that were changed since last haplotyping
            function alertUserOfUpdates() {
                if(changes.length > 0) {
                    var updateAlert = $("#update-alert");

                    var strainsList = "";
                    changes.forEach(function(strain) {
                        strainsList += ("<i><a href='" + strainMap[strain].url + "'>" + strain + "</a></i><br/>")
                    });
                    strainsList += "<br/>";

                    updateAlert.html(
                        '<div class="alert alert-info fade in"> ' +
                            '<button type="button" class="close" data-dismiss="alert" aria-label="Close">\n' +
                            '    <span aria-hidden="true">&times;</span>\n' +
                            '</button>' +
                            '<h3>Just so you know</h3>' +
                            '<hr>' +
                            '<p>It appears that since the last time haplotype probabilities were calculated on this ' +
                            'sample, there have been changes in the following contributing strain sample(s)' +
                            ':<br/><br/>' +
                            strainsList +
                            'These changes may or may not affect the haplotype probabilities if you re-run the ' +
                            'haplotype calculations at this time.<br/><br/>' +
                        '</div>');

                    updateAlert.animate({bottom: "5%"}, 300);
                }
            }

            // iterates through the contributing strains and checks for newer last_update
            function checkForUpdatesSinceHaplotyping() {
                var contrStrains = Object.keys(strainTS);

                var refTS = new Date(sample.last_haplotyping);

                contrStrains.forEach(function(strain) {
                    if(strainTS[strain]) {
                        var compTS = new Date(strainTS[strain]);

                        if(refTS < compTS) {
                            changes.push(strain);
                        }
                    }
                });

                alertUserOfUpdates();
            }

            // only check contributing strains for updates if the sample isn't a candidate
            if(!sample.haplotype_candidate) {
                getLastContributingCandidateUpdates();
                setTimeout(checkForUpdatesSinceHaplotyping, 6000);
            }

            function updateHaplotypeStrains(newContribStrains) {
                sample['contributing_strains'] = discardDupElems(newContribStrains);
                var sortedContribStrains = sample['contributing_strains'].concat().sort();

                hapKaryoPlot.drawLegend(strainMap, sortedContribStrains, 900);
                hapIntervalPlot.drawLegend(strainMap, sortedContribStrains, 150);

                readonlyHaplotypesPanel.empty();
                sortedContribStrains.forEach(function(strain, i) {
                    if(i >= 1) {
                        readonlyHaplotypesPanel.append(document.createTextNode(', '));
                    }
                    var strainIdx = strainNames.indexOf(strain);

                    var haploLink = $(
                            '<a href="' + strainMap[strain].url + '" class="hap hap' + strainIdx + '">'
                            + '<span id="sample-color-icon" class="fa fa-square" style="color: ' + strainMap[strain].color + ';"></span>'
                            + escapeHtml(strain)
                            + '</a>');

                    readonlyHaplotypesPanel.append(haploLink);
                });
                updateHaplotypeLinkHoverListeners();
                contribStrainTokensInput.tokenfield('setTokens', sample['contributing_strains']);
            }

            /**
             * creates the list of haplo links and svg legend based on the contributing
             * strains for the comparison sample
             *
             * @param contribStrains - list of all contributing strains for the comparison sample
             */
            function updateCompHaplotypeStrains(contribStrains) {
                var sortedContribStrains = contribStrains.concat().sort();

                hapIntervalCompPlot.drawLegend(strainMap, sortedContribStrains, 150);

                readonlyCompHaplotypesPanel.empty();
                sortedContribStrains.forEach(function(strain, i) {
                    if(i >= 1) {
                        readonlyCompHaplotypesPanel.append(document.createTextNode(', '));
                    }
                    var strainIdx = strainNames.indexOf(strain);

                    var haploLink = $(
                            '<a href="' + strainMap[strain].url + '" class="hap hap' + strainIdx + '">'
                            + '<span id="sample-color-icon" class="fa fa-square" style="color: ' + strainMap[strain].color + ';"></span>'
                            + escapeHtml(strain)
                            + '</a>');

                    readonlyCompHaplotypesPanel.append(haploLink);
                    updateHaplotypeLinkHoverListeners();
                });
            }

            function updateSampleID(newSampleID) {
                sampleID = newSampleID;

                var readonlySampleIDText = newSampleID;
                if(sample.other_ids) {
                    readonlySampleIDText += ' (' + sample.other_ids.join(', ') + ')';
                }

                $('[data-sampleid]').text(readonlySampleIDText);
                editSampleIDInput.val(newSampleID);
            }

            function switchEditSampleView(isEdit) {
                $('[data-display-when="sample-edit"]').css('display', isEdit ? '' : 'none');
                $('[data-display-when="sample-readonly"]').css('display', isEdit ? 'none' : '');
            }
            editSampleButton.click(function() {switchEditSampleView(true);});

            function resetEdits() {
                // restore old values to inputs so that if the user clicks edit again they
                // won't be presented with the values that they just cancelled.
                //TODO: use js var
                $('#edit-owner-input').val('{{ sample.owner }}');
                updateSampleTags(sample.tags);
                updateStandardDesignation(sample.standard_designation);
                updateHaplotypeStrains(sample.contributing_strains);
                updateSampleID(sampleID);
                updateNotes(sample.notes);
                updateSampleIsPublic(sample.is_public);
                updateSex(sex);
                updateCtrlEngTgts(sample.pos_ctrl_eng_tgts, true);
                updateCtrlEngTgts(sample.neg_ctrl_eng_tgts, false);
            }
            cancelSampleButton.click(function() {
                switchEditSampleView(false);
                resetEdits();
            });

            runHaplotypingButton.click(function() {
                runHaplotypingButton
                    .attr("disabled", true)
                    .html("<i class='fa fa-circle-o-notch fa-spin' aria-hidden='true'></i> Running...");

                var postData = {};

                postData['contributing_strains'] = JSON.stringify(sample.contributing_strains);
                postData['last_haplotyping'] = true;

                saveTagRequest = $.ajax({
                    type: "POST",
                    url: "{{ url_for('update_sample', mongo_id=sample._id) }}",
                    data: postData,
                    dataType: 'json',
                    success: function(data) {
                        var updateTs = data.ts + ' by {{ g.user.email_address }}';
                        $('#lastupdate').text(updateTs);

                        // This time stamp will mirror the last update.
                        // On reload, a more indepth time stamp will replace it.
                        $("#lasthaplotype").html(data.ts
                            + '<button id="run-haplotyping" type="button" class="btn btn-info btn-sm">'
                            + '<span class="glyphicon glyphicon-repeat" aria-hidden="true"></span> '
                            + 'Recalculate Haplotypes </button>');
                        updateHaplotypeStrains(sample.contributing_strains);
                        updateHaplotypeSVG();
                    }
                }).fail(function() {
                    showErrorMessage('An error occurred while saving sample.');
                }).always(function() {
                    saveTagRequest = null;
                    sampleUpdateSpinner.css('display', 'none');
                });

                // keep button disabled for 30 seconds to keep users from sending multiple requests
                setTimeout(function() {
                    runHaplotypingButton
                        .attr("disabled", false)
                        .html("<span class='glyphicon glyphicon-repeat' aria-hidden='true'></span> Recalculate Haplotypes");
                }, 30000)
            });

            var saveTagRequest = null;
            saveSampleButton.click(function() {
                if(saveTagRequest !== null) {
                    saveTagRequest.abort();
                    saveTagRequest = null;
                    sampleUpdateSpinner.css('display', 'none');
                }

                var postData = {};

                var newPosCtrlEngTgts = editPosEngTgtsInput.tokenfield('getTokens').map(function(x) {
                    return x.value;
                });
                newPosCtrlEngTgts = discardDupElems(newPosCtrlEngTgts);
                var posCtrlEngTgtsChanged = !arraysEq(newPosCtrlEngTgts, sample.pos_ctrl_eng_tgts);
                if(posCtrlEngTgtsChanged) {
                    postData['pos_ctrl_eng_tgts'] = JSON.stringify(newPosCtrlEngTgts);
                }

                var newNegCtrlEngTgts = editNegEngTgtsInput.tokenfield('getTokens').map(function(x) {
                    return x.value;
                });
                newNegCtrlEngTgts = discardDupElems(newNegCtrlEngTgts);
                var negCtrlEngTgtsChanged = !arraysEq(newNegCtrlEngTgts, sample.neg_ctrl_eng_tgts);
                if(negCtrlEngTgtsChanged) {
                    postData['neg_ctrl_eng_tgts'] = JSON.stringify(newNegCtrlEngTgts);
                }

                var newTags = discardDupElems(editTagsInput.tokenfield('getTokens').map(function(x) {
                    return x.value;
                }));
                var tagsChanged = !arraysEq(newTags, sample.tags);
                if(tagsChanged) {
                    postData['tags'] = JSON.stringify(newTags);
                }

                var newContribStrainTokens = contribStrainTokensInput.tokenfield('getTokens');
                var newContribStrains = discardDupElems(newContribStrainTokens.map(function(tok) {return tok.value}));
                newContribStrains.sort();
                var contribStrainsChanged = !arraysEq(newContribStrains, sample.contributing_strains);
                var haplotyping_timestamp = new Date(Date.now());
                if(contribStrainsChanged) {
                    postData['contributing_strains'] = JSON.stringify(newContribStrains);
                    postData['last_haplotyping'] = haplotyping_timestamp;
                }

                var newSampleID = editSampleIDInput.val();
                var sampleIDChanged = newSampleID !== sampleID;
                if(sampleIDChanged) {
                    postData['sample_id'] = newSampleID;
                }

                var owner = $('#edit-owner-input').val();
                postData['owner'] = owner;

                var newStrainId = editStrainIdInput.val();
                var StrainIdChanged = newStrainId !== strainID;
                if(StrainIdChanged) {
                    postData['strain_id'] = newStrainId;
                }

                var newSex = editSexSelect.val();
                var sexChanged = newSex !== sex;
                if(sexChanged) {
                    postData['sex'] = newSex;
                }

                var hapCandidate = $('#hap-candidate-checkbox').prop('checked');
                var hcChanged = hapCandidate !== sample.haplotype_candidate;
                if(hcChanged) {
                    postData['hap_cand'] = hapCandidate;
                }

                var newStandardDesignation = editStandardDesignationInput.val();
                var standardDesignationChanged = newStandardDesignation != sample.standard_designation;
                if(standardDesignationChanged) {
                    postData['standard_designation'] = newStandardDesignation;
                }

                /**
                  *  if hap cand true:
                  *  noOtherSamplesHapCandbyStrainName (ajax - need two endpoints?)
                  *  TODO: need a way to put messages below edit button
                  *  TODO: prevent form submit if conditions fail
                  */

                var cancelUpdate = false;

                if(hapCandidate==true) {

                    var msg = '';

                    //make sure there is at least one, and only one strain name applied
                    var strain_list = newStandardDesignation.split(",");

                    if (newStandardDesignation.length == 0) {
                        msg += '<br><b> - A Haplotype candidate must have at least one strain name</b>';
                        cancelUpdate = true;
                    }
                     else {

                        if (strain_list.length > 1) {
                            msg += '<br><b> - A haplotype candidate can only have one strain. Please update your sample accordingly.</b>';
                            cancelUpdate = true;
                            //now we know we have only 1 strain name applied and can run the next check
                        } else {
                            /** we have only one strain name so we can check if
                             * its been assigned to any other haplo candidate samples
                             * TODO: parse json */

                            var pdata = {}
                            pdata['strain_name'] = newStandardDesignation;
                            pdata['platform'] = '{{ sample.platform_id }}';

                            $.ajax({
                                type: "POST",
                                async: false,
                                url: "{{ url_for('check_hap_cands') }}",
                                data: pdata,
                                dataType: 'json',
                                success: function (data) {
                                    if (data.results == 'none') {
                                    } else {
                                        msg += '<br>There are other sample(s) found which are designated as haplotype ' +
                                            'candidates with this strain associated.<br>';
                                        msg += 'Click on the links below to remove the haplotype candidate designation ' +
                                            'from the samples before proceeding:<br>';

                                        $.each(data.results, function(item) {
                                            msg += '<a href="/sample/'+data.results[item].sampleId+'.html" target="blank">' +
                                                'Sample: '+data.results[item].sampleName+'</a><br>';
                                        });
                                        cancelUpdate = true;
                                        }
                                    }
                                });
                            }
                        }
                    if (newSex != 'female') {
                        msg += "<br><b> - Your sample's sex must be female in order to designate or keep it as a haplotype candidate</b>";
                        cancelUpdate = true;
                    }
                    if (cancelUpdate==true) {
                        $('#msg').html(msg);
                        $('#warning').show();
                    }

                }

                var newNotes = editNotesInput.val();
                var notesChanged = newNotes != sample.notes;
                if(notesChanged) {
                    postData['notes'] = newNotes;
                }

                var newSampleIsPublic = sampleIsPublicCheckbox.prop('checked');
                var sampleIsPublicChanged = newSampleIsPublic !== sample.is_public;
                if(sampleIsPublicChanged) {
                    postData['is_public'] = newSampleIsPublic;
                }

                if(!cancelUpdate) {
                    saveTagRequest = $.ajax({
                        type: "POST",
                        url: "{{ url_for('update_sample', mongo_id=sample._id) }}",
                        data: postData,
                        dataType: 'json',
                        success: function (data) {
                            //clear warning message and hide div
                            $('#msg').html('');
                            $('#warning').hide();
                            var updateTs = data.ts + ' by {{ g.user.email_address }}';
                            $('#lastupdate').text(updateTs);

                            $('#readonly-owner').text(owner);

                            if (tagsChanged) {
                                updateSampleTags(newTags);
                            }
                            if (posCtrlEngTgtsChanged) {
                                updateCtrlEngTgts(newPosCtrlEngTgts, true);
                            }
                            if (negCtrlEngTgtsChanged) {
                                updateCtrlEngTgts(newNegCtrlEngTgts, false);
                            }
                            if (contribStrainsChanged) {
                                updateHaplotypeStrains(newContribStrains);
                                updateHaplotypeSVG();
                            }
                            if (sampleIDChanged) {
                                updateSampleID(newSampleID);
                            }
                            if (hcChanged) {
                                $('#hap-candidate').text(hapCandidate);
                                $('#hap-candidate-checkbox').prop('checked', hapCandidate);
                            }
                            if (standardDesignationChanged) {
                                updateStandardDesignation(newStandardDesignation);
                            }
                            if (notesChanged) {
                                updateNotes(newNotes);
                            }
                            if (sampleIsPublicChanged) {
                                updateSampleIsPublic(newSampleIsPublic);
                            }
                            if (StrainIdChanged) {
                                updateSampleStrainId(newStrainId);
                            }
                            if (sexChanged) {
                                updateSex(newSex);
                                updateHaplotypeSVG();
                            }

                        }
                    }).fail(function () {
                        showErrorMessage('An error occurred while saving sample.');
                    }).always(function () {
                        saveTagRequest = null;
                        sampleUpdateSpinner.css('display', 'none');
                    });

                }

                switchEditSampleView(false);
            });

            var svg = d3.select('#hapkaryoplot-svg');

            var hapKaryoPlot = new HaploKaryoPlot({
                name: "karyoplot",
                svg: svg,
                width: 900,
                height: 900,
                chrSizes: mm10ChrSizes
            });

            hapKaryoPlot.zoomIntervalChange = updateInterval;

            var intervalSVG = d3.select("#hapintervalplot-svg");

            var hapIntervalPlot = new HaploKaryoPlot({
                name: "interval-reference",
                svg: intervalSVG,
                width: 900,
                height: 150,
                chrSizes: mm10ChrSizes,
                intervalMode: true
            });

            var intervalCompSVG = d3.select("#hapintervalplot-comp-svg");

            var hapIntervalCompPlot = new HaploKaryoPlot({
                name: "interval-comparison",
                svg: intervalCompSVG,
                width: 900,
                height: 150,
                chrSizes: mm10ChrSizes,
                intervalMode: true
            });

            hapIntervalPlot.zoomIntervalChange = updateInterval;
            hapIntervalCompPlot.zoomIntervalChange = updateInterval;

            function highlightHaplotype(strainName) {
                clearHaplotypeHightlight();
                var strainIdx = strainNames.indexOf(strainName);

                svg.selectAll('rect.hap').style('fill-opacity', 0.25);
                svg.selectAll('rect.hap' + strainIdx).style({
                    'fill-opacity': null,
                    'stroke': '#000000'
                });
                intervalSVG.selectAll('rect.hap').style('fill-opacity', 0.25);
                intervalSVG.selectAll('rect.hap' + strainIdx).style({
                    'fill-opacity': null,
                    'stroke': '#000000'
                });
                intervalCompSVG.selectAll('rect.hap').style('fill-opacity', 0.25);
                intervalCompSVG.selectAll('rect.hap' + strainIdx).style({
                    'fill-opacity': null,
                    'stroke': '#000000'
                });
                $('a.hap' + strainIdx).css({
                    'box-shadow': '0px 0px 15px 2px ' + strainMap[strainName].color
                });
            }

            function clearHaplotypeHightlight() {
                svg.selectAll('rect.hap').style({
                    'fill-opacity': null,
                    'stroke': null
                });
                intervalSVG.selectAll('rect.hap').style({
                    'fill-opacity': null,
                    'stroke': null
                });
                intervalCompSVG.selectAll('rect.hap').style({
                    'fill-opacity': null,
                    'stroke': null
                });
                $('a.hap').css({
                    'box-shadow': ''
                });
            }

            function updateHaplotypeLinkHoverListeners() {
                strainNames.forEach(function(strainName, i) {
                    $('a.hap' + i).hover(
                        function() {
                            highlightHaplotype(strainName);
                        },
                        function() {
                            clearHaplotypeHightlight(strainName);
                        }
                    );
                });
            }

            //looks like this is where the hover events for the highlight haplotype plot
            // are being registered
            hapKaryoPlot.mouseOverHaplotype = highlightHaplotype;
            hapKaryoPlot.mouseOutHaplotype = clearHaplotypeHightlight;

            var haploSnpConcordance = $('#haplo-snp-concordance');
            var updateHaplotypeSVGTimeoutID = null;
            var hkPlot = $('#hapkaryoplot');
            var hkpOverlay = new ElemOverlay(hkPlot, $('<span><span class="fa fa-refresh fa-spin" aria-hidden="true"> \
                </span> Generating Genome Karyotype Plot</span>'));

            function updateHaplotypeSVG() {
                hkpOverlay.overlayActive(true);
                $.getJSON("{{ url_for('viterbi_haplotypes_json', mongo_id=sample._id) }}", function(data) {
                    hapKaryoPlot.updateHaplotypes(data, strainMap, strainNames);
                    hapIntervalPlot.updateHaplotypes(data, strainMap, strainNames);
                    var concordantPercent = data.viterbi_haplotypes.concordant_percent;
                    haploSnpConcordance.text(
                            concordantPercent === null ? 'N/A' : concordantPercent.toFixed(2) + '%');

                    if(anyResultsPending(data)) {
                        if(updateHaplotypeSVGTimeoutID !== null) {
                            // to prevent a pileup of timeout events clear any existing timeout
                            window.clearTimeout(updateHaplotypeSVGTimeoutID);
                        }
                        updateHaplotypeSVGTimeoutID = window.setTimeout(updateHaplotypeSVG, haplotypePollingIntervalMillisecs);
                    }
                }).fail(function() {
                    showErrorMessage('An error occurred while updating haplotypes.');
                }).always(function () {
                    hkpOverlay.overlayActive(false);
                });
            }

            /**
             * gathers data for the comparison sample svg
             * @param compId - sample id of the selected comparison sample to view
             * @param attempt - number of times to attempt data retrieval before it gives up
             */
            function updateComparisonSVG(compId, attempt) {
                attempt--;

                // if user is not logged in, don't make request
                if (typeof compId === "undefined") {
                    return;
                }

                var haplotypesURL = "/sample/" + compId + "/viterbi-haplotypes.json";
                $.getJSON(haplotypesURL, function(data) {
                    hapIntervalCompPlot.updateHaplotypes(data, strainMap, strainNames);
                    updateCompHaplotypeStrains(data.contributing_strains);

                    // if data is still pending and we still have attempts left,
                    // try getting data again
                    if(anyResultsPending(data) && attempt > 0) {
                        if(updateHaplotypeSVGTimeoutID !== null) {
                            // to prevent a pileup of timeout events clear any existing timeout
                            window.clearTimeout(updateHaplotypeSVGTimeoutID);
                        }
                        updateHaplotypeSVGTimeoutID = window.setTimeout(function() {
                            updateComparisonSVG(compId, attempt)
                            }, haplotypePollingIntervalMillisecs
                        );
                    }
                    // if no attempts left, inform user that no data turned up
                    else if(anyResultsPending(data) && attempt === 0) {
                        d3.select("[class=snps-interval-comparison]").selectAll("*").remove();
                        d3.select("[class=no-data-overlay]").remove();
                        hapIntervalCompPlot.drawOverlay("no data to show", 225);
                    }
                }).fail(function() {
                    showErrorMessage('An error occurred while updating haplotypes.');
                });
            }

            var alertDiv = $("#alertDiv");

            /**
             * retrieves data on SNPs within a given sample and chromosome
             *
             * @param sampleId - the sample to search for
             * @param chr - the chromosome to search for
             * @param plotToUpdate - the plot that needs to receive the SNP data
             */
            function getSNPData(sampleId, chr, plotToUpdate) {

                // if user is not logged in, don't make request
                if (typeof sampleId === "undefined") {
                    return;
                }

                var snpsURL = "/snps.json/" + sampleId + "/" + chr;
                $.getJSON(snpsURL, function(data) {
                    plotToUpdate.updateSNPBar(data);
                }).fail(function() {
                    if (plotToUpdate == hapIntervalCompPlot) {
                        d3.select("[class=snps-interval-comparison]").selectAll("*").remove();
                    }

                    alertDiv.html('<div class="alert alert-warning fade in"> ' +
                         '<button type="button" class="close" data-dismiss="alert" aria-hidden="true">x' +
                         '</button>An error occurred while gathering SNP data.</div>');
                    alertDiv.animate({top: "90%"}, 300);
                });
            }

            alertDiv.on("click", function() {
                alertDiv.animate({top: "100%"}, 300);
            });

            switchEditSampleView(false);

            // initialize the UI
            resetEdits();
            updateHaplotypeSVG();

            // render snps
            getSNPData(sample._id, 1, hapIntervalPlot);

            // update the interval with a default interval to activate candidate haplotype table and external link buttons
            updateInterval({chr: "1", startPos: 0, endPos: 10000000, size: 10000000});
        });
    </script>
{% endblock %}
{% block content %}
    <div class="page-header">
        <h1>Sample: <span data-sampleid>{{ sample.sample_id }}</span></h1>
    </div>
    {% if g.user %}
    <div style="margin-bottom: 10px;">
        <!-- only allow admin or owner to edit a sample -->
        {% if (g.user.email_address == sample.owner) or (g.user.administrator == True) %}
        <span data-display-when="sample-readonly">
            <button id="edit-sample-button" type="button" class="btn btn-primary">
                <span class="glyphicon glyphicon-edit" aria-hidden="true"></span> Edit Sample
            </button>
            <div class="alert alert-danger" id="warning" style="display: none;">
                    <a href="#" class="close" data-dismiss="alert" aria-label="close" title="close">×</a>
                    <strong>Warning!</strong>
                    <div id="msg"></div>
                </div>
            <span id="sample-update-spinner" class="fa fa-spinner fa-pulse" style="display: none;"></span>
        </span>
        {% endif %}
        <span data-display-when="sample-edit" style="display: none;">
            <button id="save-sample-button" type="button" class="btn btn-primary">
                <span class="glyphicon glyphicon-ok" aria-hidden="true"></span> Save Edits
            </button>
            <button id="cancel-sample-button" type="button" class="btn btn-default">
                <span class="glyphicon glyphicon-remove" aria-hidden="true"></span> Cancel Edits
            </button>
        </span>
    </div>
    {% endif %}
    <div class="row">
        <div class="col-md-12">
            <table class="table">
                <tbody>
                    <tr>
                        <th>
                            Sample ID
                            <span
                                    class="glyphicon glyphicon-question-sign help-icon"
                                    aria-hidden="true"
                                    data-toggle="tooltip"
                                    data-placement="top"
                                    title="A unique ID randomly generated by HaploQA. The sample ID from the sample
                                    map file is provided in parentheses. This can be edited to any unique ID by the owner">
                            </span>
                            <b data-display-when="sample-edit" class="required"> *</b>
                        </th>
                        <td>
                            <span class="form-group" data-display-when="sample-edit" style="display: none;">
                                <input id="edit-sampleid-input" type="text" class="form-control"/>
                            </span>
                            <span data-sampleid data-display-when="sample-readonly">
                                {{ sample.sample_id }}
                                {% if sample.other_ids %}
                                    ({% for other_id in sample.other_ids %}{{ other_id }}{% if not loop.last %}, {% endif %}{% endfor %})
                                {% endif %}
                            </span>
                        </td>
                    </tr>
                    <tr>
                        <th>Owner</th>
                        <td>
                            {% if g.user.administrator == True %}
                            <span class="form-group" data-display-when="sample-edit" style="display: none;">
                                <input id="edit-owner-input" type="text" class="form-control" value="{{ sample.owner }}" placeholder="1 owner only"/>
                            </span>
                            {% endif %}
                            <span id="readonly-owner" data-display-when="sample-readonly">{{ sample.owner }}</span>
                        </td>
                    </tr>
                    {% if g.user %}
                        <tr>
                            <th>
                                Last Update
                            </th>
                            <td id="lastupdate">
                                {% if sample.last_update %}{{ sample.last_update }} by {{ sample.updated_by }}{% else %}N/A{% endif %}
                            </td>
                        </tr>
                        {% if not sample.haplotype_candidate %}
                            <tr>
                                <th>
                                    Last Haplotype Calculation
                                    <span
                                            class="glyphicon glyphicon-question-sign help-icon"
                                            aria-hidden="true"
                                            data-toggle="tooltip"
                                            data-placement="top"
                                            title="The timestamp for the last time that haplotypes were calculated
                                            for this sample">
                                    </span>
                                </th>
                                <td id="lasthaplotype">
                                    {% if sample.last_haplotyping %}{{ sample.last_haplotyping }}{% else %}N/A{% endif %}
                                    <button id="run-haplotyping" type="button" class="btn btn-info btn-sm">
                                        <span class="glyphicon glyphicon-repeat" aria-hidden="true"></span> Recalculate Haplotypes
                                    </button>
                                </td>
                            </tr>
                        {% endif %}
                    {% endif %}
                    <tr>
                        <th>
                            Strain Name(s)
                            <span
                                    class="glyphicon glyphicon-question-sign help-icon"
                                    aria-hidden="true"
                                    data-toggle="tooltip"
                                    data-placement="top"
                                    data-html="true"
                                    title="Where possible, should be used for strains that are archived by a public
                                    repository (The Jackson Laboratory, MMRRC, EMMA, etc.). The official strain name
                                    can be found on the strain data sheet/page. <b>More information can be found on the
                                    Help & About page.</b>">
                            </span>
                        </th>
                        <td>
                            <span class="form-group" data-display-when="sample-edit" style="display: none;">
                                <input id="edit-standarddesignation-input" type="text" class="form-control" value="{{ sample.standard_designation }}"/>
                            </span>
                            <span id="readonly-standarddesignation" data-display-when="sample-readonly">{{ sample.standard_designation }}</span>
                        </td>
                    </tr>
                    <tr>
                        <th>
                            Strain ID
                            <span
                                    class="glyphicon glyphicon-question-sign help-icon"
                                    aria-hidden="true"
                                    data-toggle="tooltip"
                                    data-placement="top"
                                    data-html="true"
                                    title="Refers to the unique identifier assigned to a strain by a public repository.
                                    Examples include 002020 (JAX stock number), or EM:07481 (EMMA ID), or 036938-UCD
                                    (MMRRC). If the strain doesn’t exist in a public repository, this field can be left
                                    blank or a substitute Strain ID can be used. <i>The Strain ID can be bulk edited in
                                    the previous page for any samples that share a tag.</i>">
                            </span>
                        </th>
                        <td>
                            <span class="form-group" data-display-when="sample-edit" style="display: none;">
                                <input id="edit-strain-id-input" type="text" class="form-control" value="{{ sample.strain_id }}"/>
                            </span>
                            <span id="readonly-strain-id" data-display-when="sample-readonly">{{ sample.strain_id }}</span>
                        </td>
                    </tr>
                    <tr>
                        <th>
                            Sex
                            <span
                                    class="glyphicon glyphicon-question-sign help-icon"
                                    aria-hidden="true"
                                    data-toggle="tooltip"
                                    data-placement="top"
                                    title="The sex of the animal from which the sample was collected can be provided
                                    here. Note that the inbred strain standard samples that are used for haplotype
                                    calculations are always female">
                            </span>
                        </th>
                        <td>
                            <span class="form-group" data-display-when="sample-edit" style="display: none;">
                                <select id="edit-sex-select" type="text" class="form-control">
                                    <option value="female">Female</option>
                                    <option value="male">Male</option>
                                    <option value="unknown">Unknown</option>
                                </select>
                            </span>
                            <span id="readonly-sex" data-display-when="sample-readonly">{{ sample.sex or 'Unknown' }}</span>
                        </td>
                    </tr>
                    <tr>
                        <th>
                            Platform
                            <span
                                    class="glyphicon glyphicon-question-sign help-icon"
                                    aria-hidden="true"
                                    data-toggle="tooltip"
                                    data-placement="top"
                                    data-html="true"
                                    title="The current version of HaploQA uses either MegaMUGA or GigaMUGA genotyping
                                    array platforms. <b>More information can be found on the Help & About page.</b>">
                            </span>
                        </th>
                        <td>{{ sample.platform_id }}</td>
                    </tr>
                    <tr>
                        <th>
                            % Het. Calls
                            <span
                                    class="glyphicon glyphicon-question-sign help-icon"
                                    aria-hidden="true"
                                    data-toggle="tooltip"
                                    data-placement="top"
                                    title="The percentage of probes for which heterozygous genotypes were called.">
                            </span>
                        </th>
                        <td>{{ '{0:0.2f}%'.format(sample.heterozygous_percent) }}</td></tr>
                    <tr>
                        <th>
                            % Hom. Calls
                            <span
                                    class="glyphicon glyphicon-question-sign help-icon"
                                    aria-hidden="true"
                                    data-toggle="tooltip"
                                    data-placement="top"
                                    title="The percentage of probes for which homozygous genotypes were called.">
                            </span>
                        </th>
                        <td>{{ '{0:0.2f}%'.format(sample.homozygous_percent) }}</td>
                    </tr>
                    <tr>
                        <th>
                            % No Calls
                            <span
                                    class="glyphicon glyphicon-question-sign help-icon"
                                    aria-hidden="true"
                                    data-toggle="tooltip"
                                    data-placement="top"
                                    title="The percentage of probes for which genotypes calls could not be made due to
                                    probe failure.">
                            </span>
                        </th>
                        <td>{{ '{0:0.2f}%'.format(sample.no_read_percent) }}</td>
                    </tr>
                    <tr>
                        <th>
                            % Concordance
                            <span
                                    class="glyphicon glyphicon-question-sign help-icon"
                                    aria-hidden="true"
                                    data-toggle="tooltip"
                                    data-placement="top"
                                    title="The percentage of genotype calls that are concordant with the expected,
                                    user-selected contributing strain backgrounds.">
                            </span>
                        </th>
                        <td id="haplo-snp-concordance">
                            {% if sample.viterbi_haplotypes.concordant_percent == None %}
                                N/A
                            {% else %}
                                {{ '{0:0.2f}%'.format(sample.viterbi_haplotypes.concordant_percent) }}
                            {% endif %}
                        </td>
                    </tr>
                    {% if g.user %}
                    <tr>
                        <th>
                            Tags
                            <span
                                    class="glyphicon glyphicon-question-sign help-icon"
                                    aria-hidden="true"
                                    data-toggle="tooltip"
                                    data-placement="top"
                                    title="A list of relational key words or phrases that can be assigned to individual
                                    samples to facilitate queries.">
                            </span>
                        </th>
                        <td>
                            <span class="form-group" data-display-when="sample-edit" style="display: none;">
                                <input id="edit-tags-input" type="text" class="form-control"/>
                            </span>
                            <span id="tag-links" data-display-when="sample-readonly"></span>
                        </td>
                    </tr>
                    {% endif %}
                    <tr>
                        <th>
                            Haplotype Candidate
                            <span
                                    class="glyphicon glyphicon-question-sign help-icon"
                                    aria-hidden="true"
                                    data-toggle="tooltip"
                                    data-placement="top"
                                    title="If this sample is a ‘control’ sample representing a potential contributing
                                    strain background, then choosing “true” will allow it to be used as a contributing
                                    strain for haplotype construction. Currently, only female samples can serve as
                                    haplotype candidates.">
                            </span>
                        </th>
                        <td>
                            <!-- only allow admin or curator to edit haplotype candidate designation -->
                            {% if g.user.administrator == True or g.user.curator == True%}
                                <span data-display-when="sample-edit" style="display: none;">
                                    <input id="hap-candidate-checkbox" type="checkbox" name="sample-public"> Is a Haplotype Candidate
                                </span>
                            {% endif %}
                            <span id="hap-candidate" data-display-when="sample-readonly">{{ sample.haplotype_candidate }}</span>
                        </td>
                    </tr>
                    {% if g.user %}
                    <tr>
                        <th>
                            Engineered Target Positive Controls
                            <span
                                    class="glyphicon glyphicon-question-sign help-icon"
                                    aria-hidden="true"
                                    data-toggle="tooltip"
                                    data-placement="top"
                                    data-html="true"
                                    title="HaploQA can use intensity values from these probes to infer the likelihood
                                    that a genetically engineered allele is present provided that sufficient control
                                    data exist. If the sample is known to harbor a particular genetically engineered
                                    allele (for which a probe exists), then it can serve as an engineered allele
                                    positive control. <b>More information can be found on the Help & About page.</b>">
                            </span>
                        </th>
                        <td>
                            <span class="form-group" data-display-when="sample-edit" style="display: none;">
                                <input id="edit-pos-eng-tgt-input" type="text" class="form-control"/>
                            </span>
                            <span id="pos-eng-tgt-links" data-display-when="sample-readonly"></span>
                        </td>
                    </tr>
                    <tr>
                        <th>
                            Engineered Target Negative Controls
                            <span
                                    class="glyphicon glyphicon-question-sign help-icon"
                                    aria-hidden="true"
                                    data-toggle="tooltip"
                                    data-placement="top"
                                    data-html="true"
                                    title="HaploQA can use intensity values from these probes to infer the likelihood
                                    that a genetically engineered allele is present provided that sufficient control
                                    data exist. If the sample is known to NOT harbor a particular genetically engineered
                                    allele (for which a probe exists), then it can serve as an engineered allele
                                    negative control. <b>More information can be found on the Help & About page.</b>">
                            </span>
                        </th>
                        <td>
                            <span class="form-group" data-display-when="sample-edit" style="display: none;">
                                <input id="edit-neg-eng-tgt-input" type="text" class="form-control"/>
                            </span>
                            <span id="neg-eng-tgt-links" data-display-when="sample-readonly"></span>
                        </td>
                    </tr>
                    <tr>
                        <th>
                            Visibility
                            <span
                                    class="glyphicon glyphicon-question-sign help-icon"
                                    aria-hidden="true"
                                    data-toggle="tooltip"
                                    data-placement="top"
                                    title="Controls how accessible the sample is to all HaploQA users.">
                            </span>
                        </th>
                        <td>
                            <span data-display-when="sample-edit" style="display: none;">
                                <input id="sample-is-public-checkbox" type="checkbox" name="sample-public"> Publicly Visible
                            </span>
                            <span id="readonly-sample-visibility" data-display-when="sample-readonly"></span>
                        </td>
                    </tr>
                    {% endif %}
                    <tr>
                        <th>
                            Notes:
                            <span
                                    class="glyphicon glyphicon-question-sign help-icon"
                                    aria-hidden="true"
                                    data-toggle="tooltip"
                                    data-placement="top"
                                    title="Free-form text notes associated with this sample."></span>
                        </th>
                        <td>
                            <span class="form-group" data-display-when="sample-edit" style="display: none;">
                                <textarea id="edit-notes-input" class="form-control" rows="4">{{ sample.notes  or ''}}</textarea>
                            </span>
                            <span id="readonly-notes" data-display-when="sample-readonly">{{ sample.notes or ''}}</span>
                        </td>
                    </tr>
                </tbody>
            </table>
        </div>
    </div>

    <div class="page-header">
        <br/>
        <h2>Genome Karyotype Plot</h2>
    </div>

    <p>
        Click any area of the karyotype plot to get a zoomed in view.
    </p>

    <p data-display-when="sample-edit" style="display: none;">
        You can modify the "Contributing Haplotype Strains" below and click "Save Edits". This will
        cause the haplotype structure of this strain to be inferred based on these settings (using a hidden markov
        model) and rendered in a plot below.
    </p>

    <div><strong>Contributing Haplotype Strains:</strong></div>

    <div class="readonly-haplotypes-panel" data-display-when="sample-readonly"></div>

    <form id="edit-haplotypes-panel" data-display-when="sample-edit" style="display: none;">
        <div class="form-group">
            <input type="text" class="form-control" id="contrib-strains-tokens"/>
        </div>
    </form>
    <div class="row" id="hapkaryoplot">
        <div class="col-md-12">
            <svg
                    id="hapkaryoplot-svg"
                    width="900" height="960"
                    class="haplotype-plot"></svg>
        </div>
    </div>
    <div class="row">
        <div class="col-md-12">
            <a class="btn btn-default" href="{{ url_for('sample_snp_report', mongo_id=sample._id) }}" download="{{ sample.sample_id|replace('/', '_')|replace('\\', '_')|replace(':', '_')|replace(';', '_') }}_snp_report.txt">
                <span class="glyphicon glyphicon-download-alt" aria-hidden="true"></span> Download SNP Level Haplotype Report
            </a>
            <a class="btn btn-default" href="{{ url_for('sample_summary_report', mongo_id=sample._id) }}" download="{{ sample.sample_id|replace('/', '_')|replace('\\', '_')|replace(':', '_')|replace(';', '_') }}_summary_report.txt">
                <span class="glyphicon glyphicon-download-alt" aria-hidden="true"></span> Download Haplotype Composition Summary
            </a>
            <a class="btn btn-default" id="export-png">
                <span class="glyphicon glyphicon-download-alt" aria-hidden="true"></span> Export Haplotype Plot to .PNG image
            </a>
        </div>
    </div>

        <div class="page-header">
            <br/>
            <h2>Genome Interval Plot</h2>
        </div>
        <div class="form-inline">
            <label for="interval-edit">Interval:</label>
            <div class="input-group">
                <input type="text" class="form-control" id="interval-edit" size="50" placeholder="Chr2:45Mb-55Mb">
                <span class="input-group-btn">
                    <button id="clear-interval-button" type="button" class="btn btn-default">
                        <span class="glyphicon glyphicon-remove" aria-hidden="true" aria-label="clear interval"></span>
                    </button>
                    <button id="update-interval-button" class="btn btn-default" type="button">
                        <span class="glyphicon glyphicon-ok" aria-hidden="true"></span> Update Interval
                    </button>
                </span>
            </div>
            <label class="comparison-only" for="comparison-sample" style="margin-left: 35px">Comparison Sample:</label>
            <select class="comparison-only" id="comparison-sample"></select>
        </div>
        <div class="row">
            <div class="col-md-12">
                <label class="sample-label">Sample: {{ sample.sample_id }}</label>
                <div class="readonly-haplotypes-panel"></div>
                <svg
                        id="hapintervalplot-svg"
                        width="900" height="200"
                        class="haplotype-plot"></svg>
            </div>
        </div>
        <div class="row comparison-only">
            <div class="col-md-12">
                <br/>
                <label id="comp-label" class="sample-label"></label>
                <div class="readonly-comp-haplotypes-panel"></div>
                <div id="comp-carousel" class="carousel" data-ride="carousel" data-interval="false">
                    <div class="carousel-inner" role="listbox">
                        <div class="carousel-item">
                            <svg
                                id="hapintervalplot-comp-svg"
                                width="900" height="200"
                                class="haplotype-plot">
                            </svg>
                        </div>
                    </div>
                    <a id="comp-prev" class="carousel-control left">
                        <span class="glyphicon glyphicon-chevron-left black"></span>
                    </a>
                    <a id="comp-next" class="carousel-control right">
                        <span class="glyphicon glyphicon-chevron-right black"></span>
                    </a>
                </div>
            </div>
        </div>
        <br/>
        <div class="form-inline">
            <a id="goto-ensembl-link" class="btn btn-default" href="#" target="_blank">
                <span class="glyphicon glyphicon-link" aria-hidden="true"></span> Go to Ensembl Browser
            </a>

            <a id="goto-ucsc-link" class="btn btn-default" href="#" target="_blank">
                <span class="glyphicon glyphicon-link" aria-hidden="true"></span> Go to UCSC Browser
            </a>
        </div>
    {% if g.user %}
        <div class="row">
            <div class="col-md-12">
                <br/>
                <h2 style="margin-top: 50px;">Sample: {{ sample.sample_id }}</h2>
                <h4>Candidate Haplotypes</h4>
                <table id="candidate-haplotype-table" class="table table-striped" style="margin-top: 10px;">
                    <thead><tr><th>Candidate Haplotype</th><th>Negative Log Likelihood</th></tr></thead>
                    <tbody id="candidate-haplotype-table-body" style="height: 100px;">
                    <tr>
                        <td colspan="2" style="text-align: center; vertical-align: middle;"><strong>Select an interval to view Candidate Haplotypes</strong></td>
                    </tr>
                    </tbody>
                </table>
            </div>
        </div>
        <!-- FIXME: Commenting GEMM related elements for now as they are not ready for release
        <div class="page-header">
            <h4>GEMM Probabilities</h4>
        </div>
        <div class="row">
            <div class="col-md-12">
                <table id="gemm-table" class="table table-striped" style="margin-top: 10px;">
                    <thead><tr><th>GEMM Target</th><th>Probability of Presence</th></tr></thead>
                    <tbody id="gemm-table-body" style="height: 100px;">
                    <tr>
                        <td colspan="2" style="text-align: center; vertical-align: middle;"><strong>No GEMM Probabilities Available</strong></td>
                    </tr>
                    </tbody>
                </table>
            </div>
        </div>
        <p>
        See the <a href="{{ url_for('gemm_intens_html', mongo_id=sample._id) }}">GEMM Intensities plot</a> for a more detailed view of the distribution of GEMM probe intensities.
        </p> -->
    {% endif %}
    <script>
    //enable the tooltip style
     $('[data-toggle="tooltip"]').tooltip();
    </script>
{% endblock %}
